### ISSUE TITLE
Sort entries in bookmarked playlist

### ISSUE THREAD

**Issue by ein-bandit** [TeamNewPipe/NewPipe#9497]
### Checklist

- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have read the [FAQ](https://newpipe.net/FAQ/) and my problem isn't listed.
- [X] I'm aware that this is a request for NewPipe itself and that requests for adding a new service need to be made at [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor/issues).
- [X] I have taken the time to fill in all the required details. I understand that the feature request will be dismissed otherwise.
- [X] This issue contains only one feature request.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Feature description

Clicking on a "bookmarked playlist" opens a list-view of all videos. Currently adding a new video appends it to the back (no sorting possible).

**Add an option for sorting the entries of the list.**
This could be added in 3-dot-menu on the top right as "Sorting" entry.

Clicking "Sorting" 3-dot-menu-entry could open a modal with radio buttons to select a sorting style
- time when new video was added to playlist (if available)
- reverse by added-time (if availavble)
- chronological alphabetical
- reverse chronological alphabetical
- show watched or partially watched first
- show unwatched first

option 1 and 2 would solve my issue already, but when on it could be easily extended with more options.

### Why do you want this feature?

I do use a custom playlist where I add videos to 'watch later'.
Since my list contains a lot of videos by now (200+ entries) I would need a way to organize better.
When I add a new video it gets appended at the end of the list. I do have to scroll down a lot to see newly added videos.

### Additional information

old feature request without template #9202

**Comment by Jared234** [comment]
I can work on this. However, I still have some questions regarding the implementation.

AFAIK there is currently no option to sort feeds within NewPipe, right? If that's the case I believe that I should implement this feature keeping in mind that we can potentially also use it for other lists of streams in NewPipe.
This would be my approach:
- I would create an enum that contains all of the possible sorting options.
- Then I would add a field to the `PlaylistEntity` that stores the respective sorting options.
- I would adjust the `getOrderedStreamsOf(playlistId)` method in the `PlaylistStreamDAO` interface, so that it returns the streams in the right order for the respective playlist.

As for sorting options that let you sort streams by the time you added them to the playlist: I could definitely implement that. However, this information is not currently stored in the database and I'm not sure if this would bloat the database too much.

@opusforlife2 Do you have an opinion on all of this?

**Comment by Jared234** [comment]
@SameenAhnaf I agree that this sorting option is not needed. Even without the existing theme, IMO this would be more of a filter option than a sort option.
@opusforlife2 I took a closer look at this PR and also tried it myself. I don't think we're trying to do the same thing. This PR focuses on extending the functionality of the search feature by providing "sort filters (and content filters) that are available by the supported platforms itself". For the bookmarked playlists, I would do something more like this (screenshot of the equivalent of YouTube):
![image](https://user-images.githubusercontent.com/26669009/213659704-7b987992-2cf5-428e-aa82-23dffba23d44.png)

Do you have an opinion [on this comment](https://github.com/TeamNewPipe/NewPipe/issues/9497#issuecomment-1382845734)?

**Comment by Jared234** [comment]
@opusforlife2 Yes, no problem. I'm definitely going to do that. 😄

**Comment by ein-bandit** [comment]
i'm not 100% sure how the bookmarked playlists are implemented now, but i assume the videos in the playlist are stored locally.
so i assume it could be a "sorting for the locally loaded videos" function?
we are talking about the same thing, are we?

**Comment by ChaosNicro** [comment]
That also had me confused. We have bookmarked online playlists that mirror a playlist on youtube but these are read-only. I think this is about saved playlists, the kind stored locally and read-write. I don't think they store an "added datetime" per video at the moment but just a flat order. While you could add that tag for new additions and save a setting per playlist to insert new videos at the right position to retain the sorting, a one-time sorting might be easier. That would basically be a button that parses through all entries once by a given video property and saves the result as a flat order. That combined with a per-playlist switch to insert new videos at the first position instead of the last might already accomplish what you are asking for, right?

**Comment by ein-bandit** [comment]
custom created playlist it is, yes.
an option for adding to front/back would be nice yes, could also be implicit based on the current sorting (no user choice necessary) - so "last sorting action" would need to be stored.
I think you can see the video list as an array and "just reinsert the single videos at different indices".
would fulfill the minimal requirements, yes.

**Comment by sosasees** [comment]
each video in a playlist would need "date added" only to view when a video was added.
just to sort by date added, an "added index" is enough:
just make the added index of each newly added video 1 higher than the current highest added index in the playlist

**Comment by gl-dev** [comment]
Please implement custom sorting of playlists in the playlist overview *and* the video sequence within a playlist. Thank you

**Comment by DrTexx** [comment]
Just throwing my hat in to say I'd love to help with this if I can in any way. Unfortunately I haven't touched Java/Kotlin much which I understand severely kneecaps my ability to contribute (sorry to be that guy), but if there's any brainstorming or UI/UX decisions to make happy to contribute my thoughts, so feel free to @ me in this thread.

I also have a particularly large "watch later" playlist in NewPipe (1,546 videos) as some others have mentioned and would love to see this. It takes about a minute or longer to scroll to the bottom of my playlist, but I'd strongly prefer not to use the official YouTube app, and although some forks of NewPipe have this playlist sorting and some other great additions, they don't tend to be as stable which is a dealbreaker for me.

NewPipe is kickass, appreciate the team and community contributors giving their time to make it great ❤️ :)

**Comment by phmarek** [comment]
Sorting a(n external) playlist by item name would help with playlists like https://www.youtube.com/watch?v=x-HsXBPWH3Y&list=PLtbMv4lXX2mvfaoOh5OUqrOYotNcONk6q (Dick van Dyke Show, Season 1), where the individual videos are not in the expected order.

Would have to be a kind of version sort (alphabetical/numerical by space-separated token), so that "_Season 1, Episode 16 ..._" comes after "_Season 1, Episode 2 ..._"

Thanks a lot!

**Comment by MRGiacalone** [comment]
Hello team,
I've been waiting for this feature for a long time, in fact I don't know if I'm the original requestor of this or a similar feature request. 

Using Youtube playlists as an example, if I add a new item within Youtube and keep it sorted by latest date, it shows latest on top on NP as well, which is very convenient as the list keeps getting longer.

I'd like to include two additional considerations:
- Keep the selected sort configuration as default for each native NP playlist.
- "Scroll to Top/Bottom" on the side scroll bar.

**Comment by playwithinfinity** [comment]
Feature description

Is it possible to filter and sort the private playlist in way and further to search within it for one list, song or artist?

Sorry, if anything went wrong, my excuse is, I'm new here and a bit lost.
Have a good day!
Why do you want this feature?

As I only use NewPipe to listen to music online, it would be very handy to get rid of the chaos in my playlists folder.
